internal/util/reftracker/reftracker.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reftracker
+
+import (
+	goerrors "errors"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/util/reftracker/errors"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/radoswrapper"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/reftype"
+	v1 "github.com/ceph/ceph-csi/internal/util/reftracker/v1"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/version"
+
+	"github.com/ceph/go-ceph/rados"
+)
+
+// reftracker is key-based implementation of a reference counter.
+//
+// Unlike integer-based counter, reftracker counts references by tracking
+// unique keys. This allows accounting in situations where idempotency must be
+// preserved. It guarantees there will be no duplicit increments or decrements
+// of the counter.
+//
+// It is stored persistently as a RADOS object, and is safe to use with
+// multiple concurrent writers, and across different nodes of a cluster.
+
+// Add atomically adds refs to reference tracker named `rtName`.
+// If the reftracker object doesn't exist yet, it is created. This is also
+// reflected in the return value.
+func Add(
+	ioctx radoswrapper.IOContextW,
+	rtName string,
```

Comment:
An example `rtName` and `refs` entries in source code comments help too.

Reply:
@humblec since `Add()` and `Remove()` manipulate state, I thought it might be useful to also document how the state changes over time, see https://github.com/ceph/ceph-csi/pull/2893/files#diff-69e3db766e37eaa1c0fef3163906ed7e6256e091a512d48dc441dc2c66347f31R42-R123

Let me know if you have any remarks on this.

Reply:
Thanks for adding that really helpful !! @gman0 :+1: